File: ben/prolog/P54-69-trees/p55.pl
% Construct completely balanced binary trees
% In a completely balanced binary tree, the following property holds for every node: The number of nodes in its left subtree and the number of nodes in its right subtree are almost equal, which means their difference is not greater than one.
% Write a predicate cbal_tree/2 to construct completely balanced binary trees for a given number of nodes. The predicate should generate all solutions via backtracking. Put the letter 'x' as information into all nodes of the tree.

cbal_tree(0, nil).
cbal_tree(1, t(x, nil, nil)).
cbal_tree(N, T) :-
    N > 1,
    N1 is N - 1, 
    cbal_tree(N1, T1), 
    extend_tree(T1, T).

extend_tree(nil, t(x, nil, nil)).
extend_tree(t(x, L0, R0), t(x, L0, R1)) :-
    nodes(L0, NL),
    nodes(R0, NR),
    NL >= NR, 
    extend_tree(R0, R1). 
extend_tree(t(x, L0, R0), t(x, L1, R0)) :-
    nodes(L0, NL),
    nodes(R0, NR),
    NR >= NL, 
    extend_tree(L0, L1). 

nodes(nil, 0).
nodes(t(_, L, R), N) :-
    nodes(L, NL), 
    nodes(R, NR),
    N is 1 + NL + NR. 

File: ben/haskell/Q54-60-binary-trees/p55.hs
{-
Construct completely balanced binary trees

In a completely balanced binary tree, the following property holds for every node: The number of nodes in its left subtree and the number of nodes in its right subtree are almost equal, which means their difference is not greater than one.

Write a function cbal-tree to construct completely balanced binary trees for a given number of nodes. The predicate should generate all solutions via backtracking. Put the letter 'x' as information into all nodes of the tree.
-} 

data Tree a = Empty | Branch a (Tree a) (Tree a)
              deriving (Show, Eq)

cbalTree :: Int -> [Tree Char]
cbalTree 0 = [Empty]
cbalTree 1 = [Branch 'x' Empty Empty]
cbalTree nb
    | mod (nb - 1) 2 == 0 = map (\x -> Branch 'x' x x) (cbalTree (div (nb - 1) 2))
    | otherwise = [Branch 'x' a b | a <- t1, b <- t2]
                  ++ [Branch 'x' b a | a <- t1, b <- t2]
        where t1 = cbalTree (div (nb - 1) 2)
              t2 = cbalTree ((div (nb - 1) 2) + 1) 
